pins/databricks/fs.py

```diff
@@ -0,0 +1,186 @@
+from fsspec import AbstractFileSystem
+
+from pins.errors import PinsError
+
+
+class DatabricksFs(AbstractFileSystem):
+    def ls(self, path, detail=False, **kwargs):
+        return _databricks_ls(path, detail)
+
+    def exists(self, path: str, **kwargs):
+        return _databricks_exists(path)
+
+    def open(self, path: str, mode: str = "rb", *args, **kwargs):
+        if mode != "rb":
+            raise NotImplementedError
+        return _databricks_open(path)
+
+    def get(self, rpath, lpath, recursive=False, **kwargs):
+        _databricks_get(self, rpath, lpath, recursive, **kwargs)
+
+    def mkdir(self, path, create_parents=True, **kwargs):
+        if not create_parents:
+            raise NotImplementedError
+        _databricks_mkdir(path)
+
+    def put(
+        self,
+        lpath,
+        rpath,
+        recursive=True,
+        maxdepth=None,
+        **kwargs,
+    ):
+        if not recursive:
+            raise NotImplementedError
+        if maxdepth is not None:
+            raise NotImplementedError
+        _databricks_put(lpath, rpath)
+
+    def rm(self, path, recursive=True, maxdepth=None) -> None:
+        if not recursive:
+            raise NotImplementedError
+        if maxdepth is not None:
+            raise NotImplementedError
+        if _databricks_exists(path):
+            _databricks_rm_dir(path)
+
+
+def _databricks_put(lpath, rpath):
+    w = WorkspaceClient()
+    path = os.path.abspath(lpath)
+    orig_path = path
+
+    def _upload_files(path):
+        contents = os.listdir(path)
+        for item in contents:
+            abs_path = os.path.join(path, item)
+            is_file = os.path.isfile(abs_path)
+            if is_file:
+                rel_path = os.path.relpath(abs_path, orig_path)
+                db_path = os.path.join(rpath, rel_path)
+                file = open(abs_path, "rb")
+                w.files.upload(db_path, BytesIO(file.read()), overwrite=True)
+            else:
+                _upload_files(abs_path)
+
+    _upload_files(path)
+
+
+def _databricks_get(board, rpath, lpath, recursive=False, **kwargs):
+    w = WorkspaceClient()
+    file_type = _databricks_is_type(rpath)
+    if file_type == "file":
+        board.fs.get(rpath, lpath, **kwargs)
+        return
+
+    def _get_files(path, recursive, **kwargs):
+        raw_contents = w.files.list_directory_contents(path)
+        contents = list(raw_contents)
+        details = list(map(_databricks_content_details, contents))
+        for item in details:
+            item_path = item.get("path")
+            if item.get("is_directory"):
+                if recursive:
+                    _get_files(item_path, recursive=recursive, **kwargs)
+            else:
+                rel_path = os.path.relpath(item_path, rpath)
+                target_path = os.path.join(lpath, rel_path)
+                board.fs.get(item_path, target_path)
+
+    _get_files(rpath, recursive, **kwargs)
+
+
+def _databricks_open(path):
```

Comment:
It might be nice to have these as methods, since they're related to the class. You can use the [`@staticmethod`](https://www.geeksforgeeks.org/python-staticmethod-function/) decorator to denote that it doesn't access instance-specific attributes

Reply:
Via 641c3ce